PlayFabClientSDK/source/PluginManager.cs

```diff
@@ -0,0 +1,44 @@
+using System.Collections.Generic;
+
+namespace PlayFab
+{
+    public class PluginManager
+    {
+        private Dictionary<PluginContract, IPlayFabPlugin> plugins = new Dictionary<PluginContract, IPlayFabPlugin>();
+
+        /// <summary>
+        /// The singleton instance of plugin manager.
+        /// </summary>
+        public static PluginManager Instance { get; private set; } = new PluginManager();
+
+        private PluginManager()
+        {
+        }
+
+        /// <summary>
+        /// Gets a plugin.
+        /// </summary>
+        /// <param name="contract">The app contract of plugin.</param>
+        /// <returns>The plugin implementation.</returns>
+        public IPlayFabPlugin GetPlugin(PluginContract contract)
+        {
+            if (!this.plugins.ContainsKey(contract))
+            {
+                // throw new ArgumentException("Plugin for contract is not set", nameof(contract));
+                return null;
+            }
+
+            return this.plugins[contract];
+        }
+
+        /// <summary>
+        /// Sets a custom plugin.
+        /// </summary>
+        /// <param name="contract">The app contract of a custom plugin.</param>
+        /// <param name="plugin">The custom plugin implementation.</param>
+        public void SetPlugin(PluginContract contract, IPlayFabPlugin plugin)
```

Comment:
Let's have an overload that takes a list of plugins.

Reply:
I added such an overload and then deleted it. Because it will look something like this:

public void SetPlugins(IEnumerable<Tuple<IPlayFabPlugin, PluginContract, string>> plugins)
{
 foreach (var plugin in plugins)
 {
   SetPlugin(plugin.Item1, plugin.Item2, plugin.Item3);
 }
}

The value of this method is not significant (it requires a collection which must be created first, and since someone needs to create and fill that collection - either iteratively or as a plain list - they could as well simply call SetPlugin for each item of that collection in the first place instead of allocating the collection itself and each of its items. It would simply be one of those methods that customers could write themselves if really needed. Also, if we decide to refactor the signature of these methods, with additional methods like this we will have more breaking changes. The public API at this early stage should probably have minimum things, those that are really required. This method falls more in a "nice to have" category, and at this moment i see more cons from it than pros. We can talk more about this offline if needed.

---
In reply to: [200513325](https://github.com/PlayFab/CSharpSDK/pull/97#discussion_r200513325) [](ancestors = 200513325)